examples/chat/src/network/events.rs

```diff
@@ -0,0 +1,115 @@
+use tracing::error;
+
+use super::*;
+
+mod dcutr;
+mod gossipsub;
+mod identify;
+mod ping;
+mod relay_client;
+
+pub trait EventHandler<E> {
+    async fn handle(&mut self, event: E);
+}
+
+impl EventLoop {
+    pub(super) async fn handle_swarm_event(&mut self, event: SwarmEvent<BehaviourEvent>) {
+        match event {
+            SwarmEvent::Behaviour(event) => match event {
+                BehaviourEvent::Identify(event) => events::EventHandler::handle(self, event).await,
+                BehaviourEvent::Gossipsub(event) => events::EventHandler::handle(self, event).await,
+                BehaviourEvent::RelayClient(event) => {
+                    events::EventHandler::handle(self, event).await
+                }
+                BehaviourEvent::Ping(event) => events::EventHandler::handle(self, event).await,
+                BehaviourEvent::Dcutr(event) => events::EventHandler::handle(self, event).await,
+            },
+            SwarmEvent::NewListenAddr {
+                listener_id,
+                address,
+            } => {
+                let local_peer_id = *self.swarm.local_peer_id();
+                if let Err(err) = self
+                    .event_sender
+                    .send(types::NetworkEvent::ListeningOn {
+                        listener_id,
+                        address: address.with(multiaddr::Protocol::P2p(local_peer_id)),
+                    })
+                    .await
+                {
+                    error!("Failed to send listening on event: {:?}", err);
+                }
+            }
+            SwarmEvent::IncomingConnection { .. } => {}
+            SwarmEvent::ConnectionEstablished {
+                peer_id, endpoint, ..
+            } => {
+                debug!(peer=%peer_id, ?endpoint, "Connection established");
+                match endpoint {
+                    libp2p::core::ConnectedPoint::Dialer { address, .. } => {
+                        let addr_meta = match MultiaddrMeta::try_from(&address) {
+                            Ok(meta) => meta,
+                            Err(e) => {
+                                error!(%e, "Failed to parse dialer address meta for established connection");
+                                return;
+                            }
+                        };
+
+                        if addr_meta.is_relayed() {
+                            return;
+                        }
```

Comment:
It's fine to resolve on the first connection, really.. since not all hole punches will be successful